### PR TITLE
add support for exporting metrics data model

### PR DESCRIPTION
**Pull request by rghetia**
This is the first PR to add support for exporting metrics. 

This will be followed by another PR to create a reader and start exporting (equivalent of registering exporter with view).

### CODE REVIEW

**Comment by rghetia** [comment]
> I thought the Stackdriver exporter would use the reader API in [census-instrumentation/opencensus-go#1049](https://github.com/census-instrumentation/opencensus-go/pull/1049)?

In a follow up PR I'll have Reader API to create a reader for SD. Reader API PR is still not merged (pending review).

File: metrics.go [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+
+var (
+	errLableExtraction       = errors.New("error extracting labels")
+	errUnspecifiedMetricKind = errors.New("metric kind is unpsecified")
+)
+
+// ExportMetrics exports OpenCensus Metrics to Stackdriver Monitoring.
+func (se *statsExporter) ExportMetrics(ctx context.Context, metrics []*metricdata.Metric) error {
+	if len(metrics) == 0 {
+		return errNilMetric
```
Comment (songy23) [inline metrics.go]: Why would this be an error? IMO empty metric list should be valid and we should just return nil here.

File: metrics.go [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+			Description: lbl.desc,
+			ValueType:   labelpb.LabelDescriptor_STRING,
+		})
+	}
+
+	// Now fill in those from the metric.
+	for _, key := range labelKeys {
+		labelDescriptors = append(labelDescriptors, &labelpb.LabelDescriptor{
+			Key:         sanitize(key),
+			Description: key,                            // TODO: [rghetia] when descriptor is available use that.
```
Comment (songy23) [inline metrics.go]: You can leave description blank for now.

File: metrics.go [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+	resource := metricRscToMpbRsc(metric.Resource)
+
+	metricName := metric.Descriptor.Name
+	metricType, _ := se.metricTypeFromProto(metricName)
+	metricLabelKeys := metric.Descriptor.LabelKeys
+	metricKind, _ := metricDescriptorTypeToMetricKind(metric)
+
+	if metricKind == googlemetricpb.MetricDescriptor_METRIC_KIND_UNSPECIFIED {
+		return nil, errUnspecifiedMetricKind
```
Comment (songy23) [inline metrics.go]: Same here, I don't think this is an error, otherwise we won't be able to handle the `Summary` type.

File: metrics.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+)
+
+// ExportMetrics exports OpenCensus Metrics to Stackdriver Monitoring.
+func (se *statsExporter) ExportMetrics(ctx context.Context, metrics []*metricdata.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		se.protoMetricsBundler.Add(metric, 1)
```
Comment (songy23) [inline metrics.go]: This should be `metricsBundler`.

File: metrics.go [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+		"contrib.go.opencensus.io/exporter/stackdriver.uploadMetrics",
+		trace.WithSampler(trace.NeverSample()),
+	)
+	defer span.End()
+
+	for _, metric := range metrics {
+		// Now create the metric descriptor remotely.
+		if err := se.createMetricDescriptorFromMetric(ctx, metric); err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
+			return err
```
Comment (songy23) [inline metrics.go]: nit: consider continuing handling rest metrics instead of dropping the whole batch (this can be in a different PR).

File: metrics.go [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+		}
+	}
+
+	var allTimeSeries []*monitoringpb.TimeSeries
+	for _, metric := range metrics {
+		tsl, err := se.metricToMpbTs(ctx, metric)
+		if err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
+			return err
```
Comment (songy23) [inline metrics.go]: Same here.

File: metrics.go [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+		ctx,
+		"contrib.go.opencensus.io/exporter/stackdriver.uploadMetrics",
+		trace.WithSampler(trace.NeverSample()),
+	)
+	defer span.End()
+
+	for _, metric := range metrics {
+		// Now create the metric descriptor remotely.
+		if err := se.createMetricDescriptorFromMetric(ctx, metric); err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
```
Comment (songy23) [inline metrics.go]: nit: s/2/trace.StatusCodeUnknown for better readability

File: metrics.go [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+			return err
+		}
+	}
+
+	var allTimeSeries []*monitoringpb.TimeSeries
+	for _, metric := range metrics {
+		tsl, err := se.metricToMpbTs(ctx, metric)
+		if err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
```
Comment (songy23) [inline metrics.go]: ditto

File: metrics.go [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+				// TODO(@rghetia): Don't fail fast here, perhaps batch errors?
+				// return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// metricToMpbTs converts a metric into a Stackdriver Monitoring v3 API CreateTimeSeriesRequest
```
Comment (songy23) [inline metrics.go]: s/a Stackdriver Monitoring v3 API CreateTimeSeriesRequest/a list of Stackdriver Monitoring v3 API TimeSeries

File: metrics.go [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+	if metricKind == googlemetricpb.MetricDescriptor_METRIC_KIND_UNSPECIFIED {
+		// ignore these Timeserieses. TODO [rghetia] log errors.
+		return nil, nil
+	}
+
+	timeSeries := make([]*monitoringpb.TimeSeries, 0, len(metric.TimeSeries))
+	for _, ts := range metric.TimeSeries {
+		sdPoints, err := se.metricTsToMpbPoint(ts, metricKind)
+		if err != nil {
+			return nil, err
```
Comment (songy23) [inline metrics.go]: Same, consider log the err and continue.

File: metrics.go [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+	typ := rs.Type
+	if typ == "" {
+		typ = "global"
+	}
+	mrsp := &monitoredrespb.MonitoredResource{
+		Type: typ,
+	}
+	if rs.Labels != nil {
+		mrsp.Labels = make(map[string]string, len(rs.Labels))
+		for k, v := range rs.Labels {
```
Comment (songy23) [inline metrics.go]: We may need to do some mappings from OC resource labels to SD resource labels after https://github.com/census-instrumentation/opencensus-specs/pull/235. You can leave a TODO here for now.

File: metrics.go [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+		tval = &monitoringpb.TypedValue{
+			Value: &monitoringpb.TypedValue_DoubleValue{
+				DoubleValue: v,
+			},
+		}
+
+	case *metricdata.Distribution:
+		dv := v
+		var mv *monitoringpb.TypedValue_DistributionValue
+		if dv != nil {
```
Comment (songy23) [inline metrics.go]: I think this check is not necessary?

File: metrics_test.go [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+			gj, wj := serializeAsJSON(got), serializeAsJSON(tt.want)
+			if gj != wj {
+				t.Errorf("#%d: Unmatched JSON\nGot:\n\t%s\nWant:\n\t%s", i, gj, wj)
+			}
+		}
+	}
+}
+
+func TestMetricDescriptorToMonitoringMetricDescriptor(t *testing.T) {
+	t.Skip("Metric schema has changed. Fix this test later.")
```
Comment (songy23) [inline metrics_test.go]: This is not applicable here.

File: metrics_test.go [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+			gj, wj := serializeAsJSON(got), serializeAsJSON(tt.want)
+			if gj != wj {
+				t.Errorf("#%d: Unmatched JSON\nGot:\n\t%s\nWant:\n\t%s", i, gj, wj)
+			}
+		}
+	}
+}
+
+func TestMetricTypeToMonitoringMetricKind(t *testing.T) {
+	t.Skip("Metric schema has changed. Fix this test later.")
```
Comment (songy23) [inline metrics_test.go]: ditto

**Comment by songy23** [comment]
And please rebase against master.

File: metrics_test.go [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+	startTimestamp := &timestamp.Timestamp{
+		Seconds: 1543160298,
+		Nanos:   100000090,
+	}
+	endTimestamp := &timestamp.Timestamp{
+		Seconds: 1543160298,
+		Nanos:   100000997,
+	}
+
+	tests := []struct {
```
Comment (songy23) [inline metrics_test.go]: nit: consider adding a test on built-in metrics.

File: metrics_test.go [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+				t.Errorf("#%d: Unmatched JSON\nGot:\n\t%s\nWant:\n\t%s", i, gj, wj)
+			}
+		}
+	}
+}
+
+func TestMetricDescriptorToMonitoringMetricDescriptor(t *testing.T) {
+	t.Skip("Metric schema has changed. Fix this test later.")
+
+	tests := []struct {
```
Comment (songy23) [inline metrics_test.go]: Same, consider adding a test on built-in metrics.

File: metrics.go (hunk 4 above)
Comment (rghetia) [reply metrics.go]: good catch. fixed it.

File: metrics.go (hunk 7 above)
Comment (rghetia) [reply metrics.go]: done.

File: metrics.go (hunk 5 above)
Comment (rghetia) [reply metrics.go]: done.

File: metrics.go (hunk 8 above)
Comment (rghetia) [reply metrics.go]: done.

File: metrics.go (hunk 6 above)
Comment (rghetia) [reply metrics.go]: done.

File: metrics.go (hunk 9 above)
Comment (rghetia) [reply metrics.go]: changed.

File: metrics.go (hunk 10 above)
Comment (rghetia) [reply metrics.go]: done.

File: metrics.go (hunk 11 above)
Comment (rghetia) [reply metrics.go]: added a TODO

File: metrics.go (hunk 12 above)
Comment (rghetia) [reply metrics.go]: removed.

File: metrics_test.go (hunk 13 above)
Comment (rghetia) [reply metrics_test.go]: copy paste error. removed it.

File: metrics_test.go (hunk 16 above)
Comment (rghetia) [reply metrics_test.go]: Let me look.

File: metrics_test.go (hunk 15 above)
Comment (rghetia) [reply metrics_test.go]: Let me look.

File: metrics_test.go (hunk 16 above)
Comment (rghetia) [reply metrics_test.go]: @songy23 is it ok if I do this in another PR? the reason is it would require to create a generic api to provide a formatter like [GetMetricType](https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/blob/81dc1cded66aa4c95ef161923b61e41e17073f89/stackdriver.go#L205) that works for any data type (view/metric/etc).

Comment (songy23) [reply metrics_test.go]: Yes, I'm good with leaving a TODO here and have a separate PR for it.

**Comment by rghetia** [comment]
@bogdandrutu would you be reviewing this?
